RecoTracker/ExoRegionalTracking/plugins/ExoRegionProducer.cc

```diff
@@ -0,0 +1,231 @@
+    const math::XYZVector x(trackCluster.vx(), trackCluster.vy(), trackCluster.vz());
+    if (minRadius_ < 0.0 || minTrackClusterRadius < 0.0 || (x - bs).rho() > minTrackClusterRadius)
+      pseudoROIs.emplace_back(trackClusters, i, rParam_);
+  }
+  if (pseudoROIs.size() > 1) {
+    ExoRegionItr secondToLast = pseudoROIs.end();
+    secondToLast--;
+    for (ExoRegionItr i = pseudoROIs.begin(); i != secondToLast; i++) {
+      ExoRegionItr j = i;
+      j++;
+      for (; j != pseudoROIs.end(); j++)
+        distances.emplace_back(i, j);
+    }
+  }
+
+  // Track clusters farther apart than 4 times rParam_ cannot wind up in the
+  // same ROI, so remove these pairs.
+  const auto pred = [&](const Distance &a) { return a.distance() > 4.0 * rParam_; };
+  distances.remove_if(pred);
+
+  // Do clustering.
+  while (!distances.empty()) {
+    const auto comp = [](const Distance &a, const Distance &b) { return a.distance() <= b.distance(); };
+    distances.sort(comp);
+    DistanceItr d = distances.begin();
+    if (d->distance() > rParam_)
+      break;
+
+    d->entities().first->merge(*d->entities().second);
+    d->entities().second->setInvalid();
+
+    const auto distancePred = [](const Distance &a) {
+      return (!a.entities().first->valid() || !a.entities().second->valid());
+    };
+    const auto pseudoROIPred = [](const ExoRegion &a) { return !a.valid(); };
+    distances.remove_if(distancePred);
+    pseudoROIs.remove_if(pseudoROIPred);
+  }
+
+  // Remove invalid ROIs.
+  const auto roiPred = [&](const ExoRegion &roi) {
+    if (!roi.valid())
+      return true;
+    const math::XYZVector x(roi.vx(), roi.vy(), roi.vz());
+    if ((minRadius_ >= 0.0 && (x - bs).rho() < minRadius_) ||
+        (genMatch_ != "None" && !roiContainsDecayVertex(x, decayVertices, roi.rParam())))
+      return true;
+    const double discriminatorValue = ((discriminatorCut_ > 0.0) ? getDiscriminatorValue(roi, *beamSpot) : 1.0);
+    if (discriminatorValue < discriminatorCut_)
+      return true;
+    return false;
+  };
+  pseudoROIs.remove_if(roiPred);
+
+  auto regionsOfInterest = make_unique<vector<reco::Vertex> >();
+  regionsOfInterest->clear();
+
+  const vector<double> error = {1.0, 0.0, 1.0, 0.0, 0.0, 1.0};
+
+  for (const auto &roi : pseudoROIs)
+    regionsOfInterest->emplace_back(reco::Vertex::Point(roi.vx(), roi.vy(), roi.vz()),
+                                    reco::Vertex::Error(error.begin(), error.end(), true, true));
+
+  event.put(move(regionsOfInterest));
+}
+
+void ExoRegionProducer::getDecayVertices(const edm::View<reco::GenParticle> &genParticles,
+                                         vector<math::XYZVector> &decayVertices) const {
+  for (const auto &genParticle : genParticles) {
+    if (genMatch_ == "HToSSTobbbb") {
+      if (genParticle.pdgId() != 5)
+        continue;
+      if (!genParticle.isHardProcess())
+        continue;
+      if (abs(genParticle.mother()->pdgId()) != 9000006)
```

Comment:
do you mind to add a comment just above this in order to help other users to identify the particle ? (it seems like a not SM ones ...but I might be wrong, I do not play with the pdgID since long time)

Reply:
Obviated by removal of `genMatch_`.